src/health.rs

```diff
@@ -66,6 +67,7 @@ impl HealthCheck {
             Ok(())
         }).map_err(|_| {});
 
-        core.run(work).unwrap();
+        try!(core.run(work));
+        Ok(())
```

Comment:
The `try!` followed by `Ok` is redundant. What the `try!` macro expands to is something like:

```
match core.run(work) {
   Err(e) => return e,
   Ok(o) => o
}
```

so you can get rid of the `Ok(())` and have `core.run(work)`. Notice there is no semi-colon. However, that will not fix the compiler error.

I originally told you the wrong return type for `Core::run`. With the version of tokio-core that weldr is running on, `core.run(work)` returns `Result<F::Item, F::Error>` which is the _Item_ and _Error_ type of the future. Because the _work_ future sent to `core.run` maps all errors to `()`, we cannot use `try!` here One thing you could do is:

```
core.run(work).map_err(|_| io::Error::new(io::ErrorKind::Other, "Core failed"))
```

The better thing to do would be to change [map_err](https://github.com/hjr3/weldr/pull/81/files#diff-78fdcf8a77a3e128e1573a29d435768dR68) at the end of the `work` future `.map_err(From::from)`. This says to take any error returned by the `timer.interval` stream and call [std::convert::From::from](https://doc.rust-lang.org/std/convert/trait.From.html#tymethod.from) on the result. You can also write this using a more explicit closure `.map_err(|e| From::from(e))`. This works because the `work` future returns a `tokio_timer::TimerError` type if there is an error. The tokio-timer crate implemented the `std::convert::From` trait for `tokio_timer::TimerError` to be converted to `io::Error`. (btw, I didn't go and read the source code for tokio-timer to figure that out. I mapped the error to `From::from` and crossed my fingers that it was already implemented. it is a common pattern).

One more error to go! When you have `try!(server.url().join(&self.uri_path));` the compiler is complaining that `server.url().join(&self.uri_path)` returns a `hyper::error::ParseError` type on error but `timer.interval(self.interval).for_each()` is expecting a `tokio_timer::TimerError`. Unlike our previous example, the hyper crate does not know how to convert `hyper::error::ParseError into `tokio_timer::TimerError`. If you look at the [TimerError](https://github.com/tokio-rs/tokio-timer/blob/f044bd2f58d1192aff774eac9ee06852fd90a5a7/src/timer.rs#L41) in source code it makes sense. There is no logical conversion of a url parsing error into a timeout error. Instead of using `try!` and bubbling up the error, we may want to just handle it here. You might think about doing something like:

```
let url = match server.url().join(&self.uri_path) {
    Ok(url) => url,
    Err(e) => {
        error!("Invalid health check url: {:?}", e);
        return Ok(());
    }
};
```

The main difference here is that we are returning the `Ok` variant instead of the `Err` variant. If we think about this, it makes sense. `timer.interval(...).for_each` is of type `Stream` from the futures crate. If we return an `Err`, then the entire stream is [halted immediately](https://github.com/alexcrichton/futures-rs/blob/5af2fc274a07dd5099d61c5147b5fd9d10488d8e/src/stream/mod.rs#L679). We probably do not want to halt the entire stream here, but instead abort this particular server. I say _probably_ because this is one of those errors that only happens if either a) the server url is not valid or b) the uri path is not valid. In either case, it is user supplied data that has to be fixed.

I did not expect my comment to get so long. I appreciate your work here. If nothing else, it has forced me to think through how these interactions need to work. You are very close to getting it working! I also agree with @yanns that maybe a custom error type might be good here. However, if you want to get this working as-is for now, we can always put in the custom error type later.